diff -ruNpw amx_orig/amxaux.c amx/amxaux.c
--- amx_orig/amxaux.c	2005-07-23 17:07:00.000000000 +0700
+++ amx/amxaux.c	2012-04-26 10:15:55.647912200 +0700
@@ -26,7 +26,7 @@
 #include "amx.h"
 #include "amxaux.h"
 
-size_t AMXAPI aux_ProgramSize(char *filename)
+size_t AMXAPI aux_ProgramSize(const char *filename)
 {
   FILE *fp;
   AMX_HEADER hdr;
@@ -41,7 +41,7 @@ size_t AMXAPI aux_ProgramSize(char *file
   return (hdr.magic==AMX_MAGIC) ? (size_t)hdr.stp : 0;
 }
 
-int AMXAPI aux_LoadProgram(AMX *amx, char *filename, void *memblock)
+int AMXAPI aux_LoadProgram(AMX *amx, const char *filename, void *memblock)
 {
   FILE *fp;
   AMX_HEADER hdr;
@@ -52,6 +52,7 @@ int AMXAPI aux_LoadProgram(AMX *amx, cha
     return AMX_ERR_NOTFOUND;
   fread(&hdr, sizeof hdr, 1, fp);
   amx_Align16(&hdr.magic);
+  amx_Align16((uint16_t *)&hdr.flags);
   amx_Align32((uint32_t *)&hdr.size);
   amx_Align32((uint32_t *)&hdr.stp);
   if (hdr.magic != AMX_MAGIC) {
@@ -135,15 +136,6 @@ static char *messages[] = {
   return messages[errnum];
 }
 
-enum {
-  CODE_SECTION,
-  DATA_SECTION,
-  HEAP_SECTION,
-  STACK_SECTION,
-  /* ----- */
-  NUM_SECTIONS
-};
-
 int AMXAPI aux_GetSection(AMX *amx, int section, cell **start, size_t *size)
 {
   AMX_HEADER *hdr;
diff -ruNpw amx_orig/amxaux.h amx/amxaux.h
--- amx_orig/amxaux.h	2005-07-23 17:07:00.000000000 +0700
+++ amx/amxaux.h	2012-04-26 10:15:55.660913900 +0700
@@ -30,9 +30,8 @@ extern  "C" {
 #endif
 
 /* loading and freeing programs */
-size_t AMXAPI aux_ProgramSize(char *filename);
-int AMXAPI aux_LoadProgram(AMX *amx, char *filename, void *memblock,
-                           int AMXAPI (*amx_Debug)(AMX*));
+size_t AMXAPI aux_ProgramSize(const char *filename);
+int AMXAPI aux_LoadProgram(AMX *amx, const char *filename, void *memblock);
 int AMXAPI aux_FreeProgram(AMX *amx);
 
 /* a readable error message from an error code */
--- Z:/samp-server/plugins/crashdetect/amx_orig/amxdbg.h	Sat Jul 23 16:07:00 2005
+++ Z:/samp-server/plugins/crashdetect/amx/amxdbg.h	Sun Jun 17 22:43:01 2012
@@ -27,6 +27,8 @@
 #ifndef AMXDBG_H_INCLUDED
 #define AMXDBG_H_INCLUDED
 
+#include <stdio.h>
+
 #ifndef AMX_H_INCLUDED
   #include "amx.h"
 #endif 
